src/Products/ZCatalog/plan.py

```diff
@@ -244,7 +244,13 @@ def make_key(self, query):
                 # repr() is an easy way to do this without imposing
                 # restrictions on the types of values.
                 key.append((name, repr(v)))
-
+        notkeys = [
+            name for name in key
+            if isinstance(query.get(name), dict) and "not" in query[name]
+        ]
+        if notkeys:
+            key = [name for name in key if name not in notkeys]
```

Comment:
@mamico If I'm reading https://github.com/zopefoundation/Products.ZCatalog/blob/master/src/Products/PluginIndexes/unindex.py#L433 correctly, I think it is possible to query a single index with both a "normal" (non-inverted) and "not" query at the same time; e.g. `searchResults(indexname={"query": "x", "not": "y"})`.

The assumption made by this line of code is that an index is either performing a normal query or a not query, but not both, and I don't think that's a valid assumption in this edge case.

Reply:
Maybe that's fine though. In that edge case we'll end up with ("indexname", "not") as the key -- so it'll be grouped together with queries that are only doing a `not` query, but at least it'll be grouped together with another less common case for the purpose of catalog planning rather than being grouped with other queries that don't intersect with the `not` results.

Reply:
Maybe, a catchall key for all operators  could be something like

```
        operatorkeys = [
            name for name in key
            if isinstance(query.get(name), dict)
        ]
        if operatorkeys:
            key = [name for name in key if name not in operatorkeys]
            key.extend([(name, tuple(sorted(query[name].keys()) for name in operatorkeys])
```

Reply:
@mamico Interesting idea.

But since indexes are pluggable and each type of index can have arbitrary operators, I'm a little hesitant about making it this generic, since there are probably situations where the operator has a significant impact on the runtime of the query and other situations where it does not. Better to only handle `not` at the moment, I think, so that we're only changing behavior of a specific case that we understand.

It also occurs to me that maybe determining the key for the queryplan should be delegated to the index implementation -- except that it's not necessarily trivial to find the index(es?) that corresponds to a particular name in the query.

Reply:
Another case that is not handled here is if there is a query using the `not` operator on an index that is being treated as a value index.

Reply:
> Another case that is not handled here is if there is a query using the `not` operator on an index that is being treated as a value index.

Nope, This is handled. Because those are string (with not and value included), not dict, after this https://github.com/zopefoundation/Products.ZCatalog/blob/master/src/Products/ZCatalog/plan.py#L246

Reply:
Ah, yes, I see what you mean.